FreeRTOS/FreeRTOS-Kernel
# Interrupt priority assert improvements for CM3/4/7

Interrupt priority assert improvements for CM3/4/7

Description
-----------
In the `ARM_CM3`, `ARM_CM4`, and `ARM_CM7` ports, change the assertion that `configMAX_SYSCALL_INTERRUPT_PRIORITY` is nonzero to account for the number of priority bits implemented by the hardware. Change these ports to also use the lowest priority for PendSV and SysTick, ignoring `configKERNEL_INTERRUPT_PRIORITY`. The ports are implemented such that this is a requirement for both interrupts, and the other Cortex-M ports already do this.

Test Steps
-----------
Attempt to set `configMAX_SYSCALL_INTERRUPT_PRIORITY` to a value less than `(1 << (8 - configPRIO_BITS))`, but greater than `0`. The existing assert will pass, as the priority is nonzero, but priorities in this range are not supported by the hardware, and the lower bits of `BASEPRI` are reserved. The effect is that disabling interrupts by raising `BASEPRI` does not work.

Likewise, setting `configKERNEL_INTERRUPT_PRIORITY` to a value other than the highest supported by the device will result in context switches not working correctly, because the ports assume that PendSV is the lowest priority exception.

With this change, the first case will now fail an assert, and the second case will result in a correctly configured interrupt priority because `configKERNEL_INTERRUPT_PRIORITY` is no longer used.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
